1단계 미션 결제, 율무(강기석) 제출합니다.

kkiseug opened the pull request:
# 05/31 리뷰 반영
## [코드 리뷰를 위한 Files changed](https://github.com/woowacourse/spring-roomescape-payment/pull/185/files/0941a5d055dc6feb6bdc8b8a57e5162f695661e0..897bd0be945085ad7a116f2a3f3171a6c464ee35)
이번에도 잘 부탁드립니다! 많은 부분을 변경하진 않았어요. 새로운 예외 처리와 새로 배운 것들을 공유했습니다!

<br />

---
#### [코드 리뷰를 위한 Files changed!!!!](https://github.com/woowacourse/spring-roomescape-payment/pull/185/files/0941a5d055dc6feb6bdc8b8a57e5162f695661e0..9c9772f5b9c7b63abd5f757abceeba580356d023)

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [프롤로그](https://prolog.techcourse.co.kr/studylogs/4428)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
```
[ ] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
[x] else 예약어를 쓰지 않는다.
[ ] 모든 원시 값과 문자열을 포장한다.
[ ] 일급 컬렉션을 쓴다.
[ ] 한 줄에 점을 하나만 찍는다.
[x] 줄여 쓰지 않는다(축약 금지).
[ ] 모든 엔티티를 작게 유지한다.
[ ] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
[ ] getter/setter/프로퍼티를 쓰지 않는다.
```

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
### 첫 외부 API 연결
처음으로 외부 API에 연결해보았습니다. 대기업의 경우 PG사 여러 개를 계약하여 각 PG사 별로 구현체와 컨트롤러 API를 만들어놓은 후 프론트에서 서버의 상태 코드에 따라 요청하는 PG사를 다르게 요청한다고 들었습니다! 그래서 이번에 Toss PG를 사용하는만큼 구현체에서 추상화 되지 않는 부분은 Toss prefix를 붙여주었어요.

### 외부 API의 에러코드 매핑
외부 API의 에러코드는 어디까지 매핑해줘야 하나요? 모든 에러 코드를 Enum으로 관리하나요?!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kiwoook committed:
chore: .gitignore 도커 추가 및 .editorconfig 추가

kiwoook committed:
feat: 예약 대기 파일 병합

kiwoook committed:
docs: 결제 요구사항 추가

kiwoook committed:
feat: 토스 결제 위젯 추가

kiwoook committed:
feat: 토스 결제 API 환경설정 추가

kiwoook committed:
feat: 결제 관련 클래스 및 예외 처리 추가

kiwoook committed:
feat: 결제 관련 예외 처리 및 응답 인터셉터 추가

kiwoook committed:
feat: 예약 생성 시 결제 요청 처리 및 테스트 추가

kiwoook committed:
feat: PaymentConfig 환경 변수 적용

kiwoook committed:
refactor: 예약 관련 테스트에서 데이터베이스 픽스처 통합

kiwoook committed:
test: PaymentClient 성공 테스트

kiwoook committed:
test: PaymentClient 성공 테스트

kiwoook committed:
test: 서버 오류 코드에 대한 예외 테스트 추가

kiwoook committed:
refactor: 이름 PaymentConfig -> TossPaymentConfig 변경

kiwoook committed:
refactor: 토스 예외 코드에 대한 서버 처리 추가

kiwoook committed:
refactor: 불필요한 라이브러리 제거 및 재시도 라이브러리 추가

kiwoook committed:
feat: 일시적 예외 처리 추가

kiwoook committed:
refactor: TOSS 결제 서비스 명시

kiwoook committed:
test: 일시적 오류 예외 테스트

kiwoook committed:
feat: 재시도 예외 처리 로직 추가

kiwoook committed:
refactor: application.yml 및 TossPaymentClientTest.java 파일 수정

kiwoook committed:
test: 테스트 프로파일 타임아웃 시간 설정 추가

kiwoook committed:
test: 타임아웃 테스트

kiwoook committed:
refactor: 토스 패키지 분리

kiwoook committed:
docs: 결제 요구사항 구현

kkiseug commented:
@CodeRabbit

skipped하면 나한테 CodeRabbit Pro를 줘야해.

kkiseug commented:
@CodeRabbit

진짜 서운
그럼 내 코드 전체적으로 확인하고 평가해봐

kkiseug commented:
@CodeRabbit

외부 API를 사용하는 부분의 코드를 자세히 보고 리뷰해줘.

kkiseug commented:
@CodeRabbit

너무 좋은 소리만 하지말고 비판적 시각으로 다시 해봐

kkiseug commented on src/test/resources/application-timeout.yml:
Timeout 테스트를 위해서 이런 yml 파일을 따로 분리하는 것에 대해서 어떻게 생각하시나요? 이게 좀 엮어있는 궁금증인데, 어차피 MockRestServiceServer에서 willReturn 문으로 직접 Timeout에 대한 예외를 반환하고 있는데요. 제 생각에 여기서 1초나 3초나 10초를 설정해도 어차피 mock이 정해진 값을 return 하기에 의미없는 yml 파일 같다는 생각도 들어요.

timeout에 대한 테스트를 어떤 방식으로 할 수 있을까요? 해당 부분에 대한 학습과 mock에 대한 학습이 더 필요하다고 생각이 드네요. 그리고, 근본적으로 Timeout에 대한 테스트를 진행해야 할까?를 좀 더 고민해봐도 좋을 것 같네요..

궁금한 점을 적으려 했는데 뭔가 자아성찰이 된 기분이네요. 다시 돌아와서,

> Timeout 테스트를 위해서 이런 yml 파일을 따로 분리하는 것에 대해서 어떻게 생각하시나요?

앨리는 어떻게 생각하시나요?

yebink replied:
흥미로운 방법이지만 개인적으로는 유지보수 비용에 대한 부담이 더 커 보입니다..!

현재 `application-timeout.yml` 에서 timeout 설정만 가지고 있는 것이 아니라 다른 spring 등의 설정도 함께 복사되었는데요,
spring 설정에 변경이 필요하면 관련 내용이 있는 yml 파일을 모두 찾아서 수정해주어야 합니다.
`mock이 정해진 값을 return 하기에 의미없는 yml 파일 같다` 는 부분도 맞고요 😄 

> 근본적으로 Timeout에 대한 테스트를 진행해야 할까?

이 부분은 크루들과 이야기해보면 좋을 주제네요 💬

kkiseug replied:
> spring 설정에 변경이 필요하면 관련 내용이 있는 yml 파일을 모두 찾아서 수정해주어야 합니다.
> mock이 정해진 값을 return 하기에 의미없는 yml 파일 같다 는 부분도 맞고요 😄

해당 두 부분에 대해서 답을 듣고나니 의미없는 테스트 였다는 생각이 맞다고 판단했습니다! 그래서 테스트를 지웠어요. Timeout에 대한 테스트를 진행해야 하는지는 좀 더 이야기 해보고 제 기준을 잡아가야겠습니다!

kkiseug reviewed (commented):
## 궁금한 점

yebink commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,28 @@
+
+payment:
+  token: test_gsk_docs_OaPz8L5KdmQXkzRz3y47BMw6
+  connection-timeout: 3000
```
3초로 셋팅한 이유는 무엇인가요~?

yeezy-com replied:
흠,, 크게 고민해보지 않은 부분이네요.. 페어 당시에는 사용자가 3초만 기다려도 많이 느릴 것 같다고 생각해서 3초로 설정했습니다.

질문을 받고 외부 API가 얼마나 빠를까?를 먼저 고민해봤는데요. 검색해보니 너무 다양하더라고요. 10초, 30초, 2분 혹은 아예 설정하지 않는 사람..등 그래서 그냥 토스 공식 문서에서 찾을 수 있지 않을까해서 찾아보니까 토스페이먼츠 API의 Read Timeout 시간을 30초에서 60초 사이로 하라고 안내해주고 있더라구요.

https://docs.tosspayments.com/resources/glossary/timeout#read-timeout

그래서 저는 30초에서 60초 사이의 중간값인 45초로 설정해주었습니다!!!!!

kkiseug replied:
흠,, 크게 고민해보지 않은 부분이네요.. 페어 당시에는 사용자가 3초만 기다려도 많이 느릴 것 같다고 생각해서 3초로 설정했습니다.

질문을 받고 외부 API가 얼마나 빠를까?를 먼저 고민해봤는데요. 검색해보니 너무 다양하더라고요. 10초, 30초, 2분 혹은 아예 설정하지 않는 사람..등 그래서 그냥 토스 공식 문서에서 찾을 수 있지 않을까해서 찾아보니까 토스페이먼츠 API의 Read Timeout 시간을 30초에서 60초 사이로 하라고 안내해주고 있더라구요.

https://docs.tosspayments.com/resources/glossary/timeout#read-timeout

그래서 저는 30초에서 60초 사이의 중간값인 45초로 설정해주었습니다!!!!!

추가로 해당 과정에서 Read Timeout과 Connection Timeout이라는 키워드를 얻을 수 있었고, 나름 정리해보았어용.

Connection Timeout은 클라이언트가 애초에 서버에 연결되지 못한 경우입니다. 일상으로 치면 제가 식당 앞까지만 가고 들어가지 않은 경우가 있겠죠. 혹은 잘못된 식당 주소를 보고 간 경우가 있을 것 같아요.

Read Timeout은 서버에는 요청했지만 응답이 너무 오래 오지 않는 경우입니다. 예를 들면 제가 식당에는 들어갔지만, 주문까지 했음에도 음식이 안나오는 경우라고 볼 수 있을 것 같아요.

여기서 토스는 Read Timeout에 대한 요청 시간 기준만 잡아주고, Connection Timeout 시간은 별도로 기재하고 있지 않더라고요. 이는 정말 서비스 기준에 따라 다를 수 있겠다고 생각했습니다. 저는 일단 Read Timeout과 같은 시간으로 설정했어요!

그리고 추가로 Read Timeout의 경우 제 서버에서 기다리다 못해 연결을 해제한 것이니 토스 입장에서는 요청을 다 처리했는데 응답을 주지 못하는 치명적인 상황이 생길 수도 있겠다 생각했습니다. 이러한 상황을 처리하는 것도 중요할 것 같아요.

yebink commented on src/main/java/roomescape/payment/toss/service/TossPaymentService.java:
```diff
@@ -0,0 +1,21 @@
+
+    private final TossPaymentClient tossPaymentClient;
+
+    @Retryable(retryFor = {PaymentTemporaryException.class}, maxAttempts = 3, backoff = @Backoff(delay = 500))
```
retry까지 구현하셨군요 👍

kkiseug replied:
retry를 현재는 애노테이션으로 구현하였는데, `do~while`로도 구현할 수 있고 또는 아예 재시도 하지 않을 수도 있겠더라구요. `do~while`은 실무에서 사용하지 않을 것 같긴하나 아예 재시도를 하지 않는 경우는 꽤 있을 것 같아요. 엘리가 봤을 때 결제 예외에 대해서 재시도 로직은 필수라고 생각하시나요?

예외에 따라 다를 것이라 생각은 합니다!!

yebink replied:
해당 부분은 저도 확실치 않아서 저희 회사 결제팀 개발자분께 물어봤는데요!
경우에 따라 다르다고 합니다.

결제 승인 응답은 실패로 떨어졌는데, 실제로는 결제가 되어버리는 경우도 있어서 결제를 처리하는 (지금의 경우 토스) 쪽에서 이중결제 방어가 되어 있으면 재시도를 하고, 
아니면 재시도 없이 실패로 처리하고 나중에 대사 작업을 통해 토스 쪽에서 환불이 되도록 하는 프로세스도 있다고 하네요.
이외에도 다양한 케이스가 있다고 합니다 😄

yebink commented on src/main/java/roomescape/payment/toss/interceptor/TossPaymentTemporaryErrorCode.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.Arrays;
+
+public enum TossPaymentTemporaryErrorCode {
```
> 외부 API의 에러코드는 어디까지 매핑해줘야 하나요? 모든 에러 코드를 Enum으로 관리하나요?!

율무는 일부만 enum으로 관리하도록 했네요!
그렇게 진행하신 이유는 무엇인가요?

yebink replied:
Temporary 와 Server 에러코드를 나누셨는데, 기준은 무엇이었나요~?

kkiseug replied:
모든 에러 코드를 매핑하게 되면 너무 외부 의존적인 예외가 생길 것 같더라고요. 또, 외부 예외 코드가 변경되면 그만큼 제 프로젝트도 변경되는 사항이 많습니다. 또한, 이를 외부 업체에서 알려주지 않으면 치명적인 상황이 발생할 수도 있다고 생각했습니다. 그래서 크루들과 이야기 하고 일부만 Enum으로 관리하는 방식이 낫겠다고 결론을 내서 이렇게 진행했습니다!

다만 아직 외부 API 에러 코드에 대한 분류 및 선별이 애매하고 모호한 부분들이 있어서 뭔가 기준을 잡기 어려웠던 것 같아요. 그래서 일단 크게 토스 서버 내부의 ServerError, 토스 결제 진행 중 발생한 ProcessError로 나누었어요. 어떻게 보면 유저에게 보여줄 메시지가 담긴 에러와 아닌 에러를 나눈 것이라고도 볼 수 있는 것 같습니다. 추후에 좀 더 살펴보고 세분화 해보자는 것이었죠. 그 과정에서 재시도가 필요한 에러 코드에 대해서는 Temporary 예외로 또 나누었습니다.

yebink commented on src/main/java/roomescape/payment/toss/config/TossPaymentConfig.java:
```diff
@@ -0,0 +1,63 @@
+import roomescape.payment.toss.service.TossPaymentClient;
+
+@Configuration
+public class TossPaymentConfig {
```
> 그래서 이번에 Toss PG를 사용하는만큼 구현체에서 추상화 되지 않는 부분은 Toss prefix를 붙여주었어요.

확장성 👍

kkiseug replied:
감사합니다~!

yebink commented on src/main/java/roomescape/payment/controller/excpetion/PaymentExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+                .body(ApiResponse.fail(paymentErrorCode));
+    }
+
+    @ExceptionHandler(PaymentProcessException.class)
+    public ResponseEntity<ApiResponse<Void>> handlePaymentProcessException(PaymentProcessException e) {
+        PaymentErrorCode paymentErrorCode = new PaymentErrorCode(PROCESS_ERROR_CODE, e.getMessage());
+
+        return ResponseEntity
+                .status(HttpStatus.BAD_REQUEST)
+                .body(ApiResponse.fail(paymentErrorCode));
+    }
```
모든 PaymentProcessException 이 정말 400 에러일까요?
4xx 에러와 5xx 에러의 차이는 무엇일까요~?

kkiseug replied:
4xx에러는 기본적으로 클라이언트의 요청에 문제가 있는 경우를 의미하고 5xx 에러는 서버에서 요청을 처리할 수 없는 상황임을 의미합니다! 모든 PaymentProcessException을 400에러로 처리하고 있지만 이렇게 처리하면 안된다고 생각합니다!

하지만, 지금까지는 서버 입장이 저였겠지만.. 외부 API 호출을 하는 코드가 생겼으므로 이때는 요청하는 입장인 제가 클라이언트가 되고, 토스가 서버가 되어 사용자에게 응답을 해야해서 헷갈리는 부분이 많은 것 같습니다. 또한, 유저에게 무엇을 보여주고 보여주지 않아야 하는지도 좀 헷갈리는 것 같아요.

그래서 추가로 NOT_FOUND에 대해서도 예외 케이스를 추가해주었어요. 403번대 에러도 결국 사용자 입장에서 400에러로 처리하면 될 것 같다고 생각했기에 404만 따로 처리해주었습니다.

yebink replied:
사실 토스에서 내려주는 예외가 워낙 깔끔하고 구체적이어서 그대로 내려도 상관 없다고 생각하는데요..!

일반적으로 외부 API 호출에 대한 실패는 500 에러로 처리하는 경향이 있습니다.

고객의 실수가 아니라 외부 API 에 문제가 있는 것이니까요 😄 

다시 말해 4xx 에러는 에러의 원인이 고객이기 때문에, 고객이 입력값을 수정해서 200 응답을 받을 수 있는 상태이지만,
5xx 에러는 그게 아니므로 개발자의 인지 및 대응이 필요합니다.
토스 API 에 문제가 있다면, 토스 측에 즉시 알리고 토스로 결제가 불가하도록 조치를 취해야겠죠?

해당 내용 다음 단계에서도 고민해보세요!

yebink reviewed (changes requested):
안녕하세요 율무~

방탈출 결제 1단계 잘 구현하셨네요! 👍 
몇 가지 피드백 남겨두었으니 확인해주세요~

kkiseug committed:
refactor: Toss 결제 Timeout 시간 변경

kkiseug committed:
chore: 테스트 환경 관련 타임아웃 설정 및 테스트 코드 제거

- application-timeout.yml 파일 삭제
- TossPaymentTimeoutTest 관련 테스트 코드 제거
- application.yml에서 payment 토큰 및 타임아웃 설정 제거

kkiseug committed:
feat: 결제 내역 조회 실패 예외 처리 추가

- PaymentNotFoundException 예외 클래스 추가
- TossPaymentNotFoundErrorCode Enum 및 관련 로직 구현

kkiseug reviewed (commented):
예외처리는 항상 어려운 것 같습니다. 좀 더 기준을 잡아가야 할 필요가 있을 것 같아요.

kkiseug commented:
@CodeRabbit

https://github.com/woowacourse/spring-roomescape-payment/pull/185/files/9c9772f5b9c7b63abd5f757abceeba580356d023..897bd0be945085ad7a116f2a3f3171a6c464ee35

해당 부분의 변경사항을 보고 `외부 API의 예외 핸들링`에 대해서 평가해주고 개선 방향이 있다면 정답을 알려주기 보다는 키워드 위주로 알려줘.

yebink commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,29 @@
+  secret-key: secretsecretsecretsecretsecretsecret
+  expire-time: 31536000000
+
+toss:
```
👍

yebink reviewed (approved):
안녕하세요 율무~

피드백 잘 반영해주셨네요!
이번 단계는 이만 머지할게요.
질문에 대한 답변 남겨두었으니 확인해주세요~

kkiseug commented:
@coderabbitai full review
